src/unix-manager.c

```diff
@@ -652,20 +656,130 @@ static TmEcode UnixManagerCaptureModeCommand(json_t *cmd,
     SCReturnInt(TM_ECODE_OK);
 }
 
-static TmEcode UnixManagerReloadRules(json_t *cmd, json_t *server_msg, void *data)
+static TmEcode UnixManagerReloadRulesWrapper(json_t *cmd, json_t *server_msg, void *data, int do_wait)
 {
     SCEnter();
-    DetectEngineReloadStart();
 
-    while (DetectEngineReloadIsDone() == 0)
-        usleep(100);
+    if (SuriHasSigFile()) {
+        json_object_set_new(server_msg, "message",
+                            json_string("Live rule reload not possible if -s "
+                                        "or -S option used at runtime."));
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    int r = DetectEngineReloadStart();
+
+    if (r == 0 && do_wait) {
+        while (DetectEngineReloadIsIdle() == 0)
```

Comment:
did you test loading a large ruleset with config settings that make loading very slow. Can you break out of this loop then? Or is suricata stuck until loading is complete.

Reply:
I tested it loading 19082 rules, if you use the command `ruleset-reload-rules`, Suricata (unix client) is blocked until loading is completed.
If you use instead the command `ruleset-reload-nonblocking` it doesn't stuck.

Reply:
My point is that if someone presses 'Ctrl-C' while the blocking load is going it might be good to break out of the loop? Did you test this?

Reply:
Actually when 'ctrl-c' is pressed, Suricata is killed after that the reload is completed.
I've added the condition below to try to break the loop, but it doesn't work as expected:
```
while (DetectEngineReloadIsIdle() == 0 && suricata_ctl_flags == 0) {
    usleep(100)
}
```

Reply:
Why not?

Reply:
What I expect with the check above is that when I press 'ctrl-c' the loop should be broken without waiting.
So in unix client I should see "done" just after the signal is sent:
```
>>> ruleset-reload-rules
Success:
"done"
```
but this occurs when all the rules are reloaded.

Reply:
Have you tried figuring out why this happens?